# Домашнє завдання до теми "Python Basics"
В цьому ДЗ важливою умовою виконання завдань є те, що ми не імпортуємо жодних бібліотек, а робимо все на чистому Python, використовуючи лише базовий функціонал. Це необхідно аби відпрацювати роботу з базовими типами даних. А використовувати бібліотеки ми ще навчимось далі :)

# Робота з рядками

1.Запишіть у змінну `name` ваше імʼя.

In [1]:
name = 'Yuliia'

2.Запишіть у змінну `surname` ваше прізвище.

In [3]:
surname = 'Korotchenko'

3.Сконкатенуйте імʼя та прізвище в одну фразу через пробіл. Запишіть результат у нову змінну `name_surname` та виведіть результат:

In [5]:
name_surname = name + " " + surname
name_surname

'Yuliia Korotchenko'

Очікувана відповідь (тільки з вашими даними): 'Hanna Pylieva' 

4.Зробіть усі літери змінної `name_surname` заглавними використовуючи вбудовану функцію Python для рядків:

In [9]:
name_surname.upper()

'YULIIA KOROTCHENKO'

Очікувана відповідь (тільки з вашими даними): 'HANNA PYLIEVA'

5.Виведіть третю літеру вашого `name_surname`. Тут уважно, треба саме третя літера. Згадайте, з якого значення починається індексування значень в Python.

In [10]:
name_surname[2]

'l'

Очікувана відповідь (тільки з вашими даними): 'n'

6.Виведіть першу та останню літеру вашого прізвища.

In [16]:
(name_surname[0], name_surname[-1])

('Y', 'o')

Очікувана відповідь (тільки з вашими даними): ('P', 'a')

7.Виведіть довжину вашого прізвища.

In [14]:
len(surname)

11

Очікувана відповідь (тільки з вашими даними): 7

8.Виведіть кількість літер `a` у столиці Мадагаскару `Antananarivo`. Важливо врахувати всі літери, як великі, так і маленькі.

In [15]:
capital = 'Antananarivo'
capital.count('A') + capital.count('a')

4

Очікувана відповідь: 4

# Робота з числами та операторами

9.Створіть змінні a, b, c та d цілочисельного типу, задайде їм значення відповідно 1, 2, 3, 4, та обчисліть з допомогою Python наступний вираз:  $$ \frac{(a+b) * c}{d} $$

In [87]:
a = int(1)
b = int(2)
c = int(3)
d = int(4)

(a + b) * c / d

2.25

Очікувана відповідь: 2.25

10.На змінних, створених в попердньому завданні, обчисліть наступний вираз з допомогою Python: $$ (a^b + c^d)^2 $$

In [88]:
((a ** b) + (c ** d)) ** 2

6724

Очікувана відповідь: 6724

# Робота зі колекціями

У нас є наступний список:

In [1]:
l = [
    78, 12, 14, 27, 47, 28, 51, 46, 94, 37, 42, 20, 52, 33, 11, 11, 51,
    20, 81, 54, 86, 51, 38, 42, 49, 99, 76, 17,  9, 23, 97, 29, 29, 63,
    40, 11, 31, 76, 15, 21, 29, 75, 35,  2, 10, 71, 17, 10, 13,  7, 44,
    12, 19, 61, 13, 25, 62, 46, 10, 44, 75, 57, 20, 57,  3, 44, 16, 28,
    79, 57, 62, 70, 95,  8, 47, 82, 80, 45, 36, 44, 79, 95, 73, 42, 29,
    83, 80, 66, 56,  8, 73,  5, 99, 65, 65, 98, 17, 36, 16, 60
]

Будемо з ним працювати в наступних завданнях.

11.Виведіть 5 перших елементів списку

In [18]:
l[:5]

[78, 12, 14, 27, 47]

Очікувана відповідь: [78, 12, 14, 27, 47]

12.Виведіть максимальний і мінімальний елемент у списку

In [46]:
print(f'Max element {max(l)}')
print(f'Min element {min(l)}')

Max element 99
Min element 2


Очікувана відповідь: 

```
Max element  99
Max element  2
```

13.Чи є в списку елементи, які повторюються? Перевірте це програмно і з допомогою інструкцій `if ... else` виведіть на екран відповідь "Так, є елементи, які повторюються", або "Ні, всі елементи унікальні".
Важливо це зробити без використання імпортів будь-яких бібліотек, на чистому Python так би мовити.

In [47]:
if len(l) != len(set(l)):
    print('Так, є елементи, які повторюються')
else:
    print('Ні, всі елементи унікальні')

Так, є елементи, які повторюються


14.Для кожного унікального елементу списку `l` знайдіть, скільки разів він повторюється та запишіть результат у змінну `n_repeats`. Виведіть `n_repeats` на екран.

In [2]:
n_repeats = {i:(l).count(i) for i in (l)}
n_repeats

{78: 1,
 12: 2,
 14: 1,
 27: 1,
 47: 2,
 28: 2,
 51: 3,
 46: 2,
 94: 1,
 37: 1,
 42: 3,
 20: 3,
 52: 1,
 33: 1,
 11: 3,
 81: 1,
 54: 1,
 86: 1,
 38: 1,
 49: 1,
 99: 2,
 76: 2,
 17: 3,
 9: 1,
 23: 1,
 97: 1,
 29: 4,
 63: 1,
 40: 1,
 31: 1,
 15: 1,
 21: 1,
 75: 2,
 35: 1,
 2: 1,
 10: 3,
 71: 1,
 13: 2,
 7: 1,
 44: 4,
 19: 1,
 61: 1,
 25: 1,
 62: 2,
 57: 3,
 3: 1,
 16: 2,
 79: 2,
 70: 1,
 95: 2,
 8: 2,
 82: 1,
 80: 2,
 45: 1,
 36: 2,
 73: 2,
 83: 1,
 66: 1,
 56: 1,
 5: 1,
 65: 2,
 98: 1,
 60: 1}

Очікуваний результат (три крапки - для скорочення, у вас буде виведений повний набір унікальних елементів): 
```
{2: 1,
 3: 1,
 5: 1,
    ...
 97: 1,
 98: 1,
 99: 2}
```

15.Знайдіть три найменші унікальні елементи в списку `l`.

In [50]:
print(sorted(set(l))[:3])
# якщо потрібно було виконати з використанням циклу, то я думала над цим, але так і не змогла вирішити як це має бути

[2, 3, 5]


Очікувана відповідь: [2, 3, 5]


16.Використовуючи `n_repeats` скажіть, скільки разів повторюється кожен з трьох найменших елементів в списку.

In [3]:
n_repeats2 = {2:n_repeats[2], 3:n_repeats[3], 5:n_repeats[5]}
# УРА! Бо я спочатку пробувала зробити це циклом, не виходило. Якщо можна циклом, то підкажіть, будь ласка, як.

Очікувана відповідь:  {2: 1, 3: 1, 5: 1}

17.Створіть новий список `l2`, який містить кожен 11-й елемент списку `l`. Виведіть `l2` на екран.  
Памʼятаємо, щоб наприклад, перший 11-и1 елемент списку має індекс 10.

In [51]:
l2 = l[10::11]
print(l2)

[42, 51, 29, 2, 13, 44, 80, 66, 16]


Очікуваний результат: `[42, 51, 29, 2, 13, 44, 80, 66, 16]`

18.Створіть новий список `l2_squared`, де кожен елемент списку `l2` є піднесеним до степіня 2. 

In [54]:
l2_squared = []
for i in (l2):
    l2_squared.append(i ** 2)
print(l2_squared)

[1764, 2601, 841, 4, 169, 1936, 6400, 4356, 256]


Очікуваний результат: `[1764, 2601, 841, 4, 169, 1936, 6400, 4356, 256]`

19.Знайдіть суму всіх елементів списку `l2` та виведіть її на екран.

In [55]:
sum(l2)

343

In [67]:
20.Знайдіть середнє значення елементів списку `l2`, збережіть результат у змінну `l2_mean` та виведіть новостворену змінну та ії тип на екран.

SyntaxError: invalid syntax (3367619299.py, line 1)

In [66]:
l2_mean = sum(l2) / len(l2)
print(int(l2_mean))
type(l2)

38


list

In [ ]:
21.Створіть новий список `l2_gt_mean`, за наступним правилом: "якщо цей елемент спиcку `l2` є більше за `l2_mean`, то запишемо 1, інакше 0. Виведіть новий список на екран.

In [84]:
# починаючи з виразу чомучь не виводився результат, погуглила, що можна спочатку ставити умову тому спробувала і так працює
l2_gt_mean = [1 if i > int((l2_mean)) else 0 for i in l2]
l2_gt_mean

[1, 1, 0, 0, 0, 1, 1, 1, 0]

Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

# Ураа!! Ви завершили перше домашнє завдання!!!  
Аби його здати - залийте ноутбук у гітхаб репозиторій, який ви створили у перші дні навчання на курсі, та додайте посилання на файл, заповнишви форму https://forms.gle/EvxoiJiBTkuxFmhV7   
Так тримати! 🚀🚀🚀
